perses/annihilation/new_relative.py

```diff
@@ -410,6 +412,50 @@ def _constraint_check(self):
                 if constraint != constraint_from_old_system:
                     raise ValueError("Constraints are changing during switching.")
 
+    def _constraint_check_fast(self):
+        """
+        This method will check for changing constraints by first serializing the new and old systems to xml, then using
+        that xml to check for constraint changes. Using lxml and XPATH, this should be considerably faster than the
+        OpenMM API. If a constraint is found to be changing, an exception will be raised, as this cannot currently be
+        handled by the HybridTopologyFactory.
+        """
+        #set up an xpath string to find constraints
+        constraint_string = '/System/Constraints/Constraint'
+
+        #get a reference to maps with shorter names
+        o_h_map = self._old_to_hybrid_map
+        n_h_map = self._new_to_hybrid_map
+
+        #serialize the systems
+        old_system_xml = openmm.XmlSerializer.serialize(self._topology_proposal.old_system)
+        new_system_xml = openmm.XmlSerializer.serialize(self._topology_proposal.new_system)
+
+        #get the serialized systems into stringio form
+        old_system_io = StringIO(old_system_xml)
+        new_system_io = StringIO(new_system_xml)
+
+        #parse the xml strings
+        old_system_tree = etree.parse(old_system_io)
+        new_system_tree = etree.parse(new_system_io)
+
+        #get the list of constraints from new and old systems:
+        old_system_constraint_list = old_system_tree.xpath(constraint_string)
+        new_system_constraint_list = new_system_tree.xpath(constraint_string)
+
+        #convert the list of constraint elements to dictionaries. By using frozenset, we can do this independent of the order of
+        old_system_constraints = {frozenset((o_h_map[int(constraint.attrib['p1'])], o_h_map[int(constraint.attrib['p2'])])) : float(constraint.attrib['d']) for constraint in old_system_constraint_list}
+        new_system_constraints = {frozenset((n_h_map[int(constraint.attrib['p1'])], n_h_map[int(constraint.attrib['p2'])])) : float(constraint.attrib['d']) for constraint in new_system_constraint_list}
```

Comment:
I think it should be safe to use simply a `tuple` instead of a `frozenset` for the keys.

Reply:
Ah! Or is it to avoid the case when you have duplicate `(p1, p2)` and `(p2, p1)` pairs?

Reply:
But if I do that, then the order of the indices will matter, right?

Reply:
Oops, didn't see your other message.